python/equistore-operations/equistore/operations/_utils.py

```diff
@@ -192,3 +275,90 @@ def _check_gradient_presence(block: TensorBlock, parameters: List[str], fname: s
                 f"requested gradient '{parameter}' in {fname} is not defined "
                 "in this tensor"
             )
+
+
+def _labels_equal(a: Labels, b: Labels, exact_order: bool) -> bool:
+    """
+    For 2 :py:class:`Labels` objects ``a`` and ``b``, returns true if they are
+    exactly equivalent in names, values, and elemental positions. Assumes that
+    the Labels are already searchable, i.e. they belong to a parent TensorBlock
+    or TensorMap.
+    """
+    # They can only be equivalent if the same length
+    if len(a) != len(b):
+        return False
+    # Check the names
+    if not np.all(a.names == b.names):
+        return False
+    if exact_order:
+        return np.all(np.array(a == b))
+    return np.all([a_i in b for a_i in a])
+
+
+def _comparison_impl(
+    operation: str, tensor_1: TensorMap, tensor_2: TensorMap, **extra_args
+) -> str:
+    """Abstract function to do a logic operation betwee two TensorMaps."""
+    message = _check_same_keys_impl(tensor_1, tensor_2, "equal")
+    if message != NO_ERROR:
+        return f"the tensor maps have different keys: {message}"
+
+    for key, block_1 in tensor_1.items():
+        message = _comparison_block_impl(
+            operation=operation,
+            block_1=block_1,
+            block_2=tensor_2.block(key),
+            **extra_args,
+        )
+        if message != NO_ERROR:
+            return f"blocks for key '{key}' are different: {message}"
+    return NO_ERROR
+
+
+def _comparison_block_impl(
+    operation: str, block_1: TensorBlock, block_2: TensorBlock, **extra_args
```

Comment:
TorchScript does not support kwargs. Also does not support lambda functions and function definitions inside functions. I gave it a thought how to get something similar as you do with the comparison function, but I could not come up with anything that is compatible with TorchScript. The problem is that you want arbitrary callables **and** arbitrary key word arguments. I could not find a solution that can do both

Reply:
Okay I can restructure this. But passing functions is fine? Then we can just factor out this logic.